compiler/ballerina-parser/src/main/java/io/ballerina/compiler/internal/parser/BallerinaParser.java

```diff
@@ -11946,27 +11956,37 @@ private boolean isNestedQueryExpr() {
         return Collections.frequency(this.errorHandler.getContextStack(), ParserRuleContext.QUERY_EXPRESSION) > 1;
     }
 
-    private boolean isValidIntermediateQueryStart(SyntaxKind syntaxKind) {
+    private boolean isValidIntermediateQueryStart(STToken token) {
+        SyntaxKind syntaxKind = token.kind;
         switch (syntaxKind) {
             case FROM_KEYWORD:
             case WHERE_KEYWORD:
             case LET_KEYWORD:
             case SELECT_KEYWORD:
-            case COLLECT_KEYWORD:
             case JOIN_KEYWORD:
             case OUTER_KEYWORD:
             case ORDER_KEYWORD:
-            case GROUP_KEYWORD:
             case BY_KEYWORD:
             case ASCENDING_KEYWORD:
             case DESCENDING_KEYWORD:
             case LIMIT_KEYWORD:
                 return true;
+            case IDENTIFIER_TOKEN:
+                return isGroupOrCollectKeyword(token);
             default:
                 return false;
         }
     }
 
+    private static boolean isGroupOrCollectKeyword(STToken nextToken) {
+        if (nextToken.kind != SyntaxKind.IDENTIFIER_TOKEN || !(nextToken instanceof STIdentifierToken)) {
```

Comment:
The only way we can have a `SyntaxKind.IDENTIFIER_TOKEN` with instanceof not being `STIdentifierToken` instance, is when we insert an identifier via the error handler. In that case, we can use `isMissing()` API to distinguish whether it is an error handler inserted identifier. 
```suggestion
        if (nextToken.kind != SyntaxKind.IDENTIFIER_TOKEN || nextToken.isMissing()) {
```